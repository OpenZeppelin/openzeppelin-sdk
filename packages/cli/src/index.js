// module information
const version = 'v' + require('../package.json').version

// model objects
import commands from './commands'
import scripts from './scripts'
import local from './models/local'
import network from './models/network'
import TestHelper from './models/TestHelper'
import runWithTruffle from './utils/runWithTruffle'

export {
  version,
  local,
  network,
  commands,
  scripts,
  runWithTruffle,
  TestHelper,
}
