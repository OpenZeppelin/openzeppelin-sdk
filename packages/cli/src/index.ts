// module information
const version = 'v' + require('../package.json').version;

// commands
import commands from './commands';
import scripts from './scripts';

// model objects
import files from './models/files';
import local from './models/local';
import network from './models/network';
import TestHelper from './models/TestHelper';
import ConfigVariablesInitializer from './models/initializer/ConfigVariablesInitializer';

// utils
import * as naming from './utils/naming';
import log, { silent } from './utils/stdout';
const stdout = { log, silent };

export {
  version,
  files,
  local,
  network,
  commands,
  scripts,
  naming,
  stdout,
  ConfigVariablesInitializer,
  TestHelper,
};
