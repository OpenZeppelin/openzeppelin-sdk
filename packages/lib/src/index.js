// module information
const version = 'v' + require('../package.json').version

// helpers
import decodeLogs from './helpers/decodeLogs'
import encodeCall from './helpers/encodeCall'

// utils
import Proxy from './utils/Proxy'
import Logger from './utils/Logger'
import FileSystem from './utils/FileSystem'
import Contracts from './utils/Contracts'
import { sendTransaction, deploy } from './utils/Transactions'

// test behaviors
import { behaviors, helpers } from './test'
const assertions = helpers.assertions
const assertRevert = helpers.assertRevert

// model objects
// import App from './app/App'
// import Package from './package/Package'
// import AppDirectory from './directory/AppDirectory'
// import ImplementationDirectory from './directory/ImplementationDirectory'
// import FreezableImplementationDirectory from './directory/FreezableImplementationDirectory'

// TODO: UPDATE me
export {
  version,
  decodeLogs,
  encodeCall,
  assertRevert,
  assertions,
  behaviors,
  sendTransaction,
  deploy,
  Proxy,
  Logger,
  FileSystem,
  Contracts,
  // App,
  // ImplementationDirectory,
  // FreezableImplementationDirectory,
  // AppDirectory,
  // Package,
}
