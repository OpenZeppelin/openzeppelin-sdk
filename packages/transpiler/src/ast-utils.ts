import find from 'lodash.find';
import Ajv from 'ajv';
import util from 'util';
import astNodeSchema from './schemas/ast-node';

let nodeSchemaValidator = new Ajv({ allErrors: true });

/**
 * Check if given argument is an AST Node
 * @param {Object} possibleObject Argument to check for validity
 * @returns {Boolean} isAValidASTNode true if given argument is an AST node, false otherwise
 */
const isASTNode = nodeSchemaValidator.compile(astNodeSchema);

// For internal use. Throws if is passed an invalid AST node, else does nothing.
function throwIfInvalidNode(node) {
  if (!isASTNode(node)) {
    throw new Error(util.inspect(node) + ' is not a valid AST node.');
  }
}

function isContractKind(node, kind) {
  throwIfInvalidNode(node);
  return node['contractKind'] === kind;
}

function isInterface(node) {
  return isContractKind(node, 'interface');
}

function isInterface(node) {
  return isContractKind(node, 'interface');
}

function isContract(node) {
  return isContractKind(node, 'contract');
}

/**
 * @param {Object} node The node to check
 * @param {String} name The type of the node
 * @returns {Boolean} true if the given node is the right type
 */
function isNodeType(node, name) {
  throwIfInvalidNode(node);
  return node['nodeType'] === name;
}

function isImportDirective(node) {
  return isNodeType(node, 'ImportDirective');
}

function isVarDeclaration(node) {
  return isNodeType(node, 'VariableDeclaration');
}

function isContractType(node) {
  return isNodeType(node, 'ContractDefinition');
}

function isPragmaDirective(node) {
  return isNodeType(node, 'PragmaDirective');
}

function idModifierInvocation(node) {
  return isNodeType(node, 'ModifierInvocation');
}

/**
 * @param {Object} node The node to check
 * @returns {Boolean} true if the given node is an BlockStatement
 */
function isBlockStatement(node) {
  return isNodeType(node, 'BlockStatement');
}

/**
 * @param {Object} node The node to check
 * @returns {Boolean} true if the given node is an BreakStatement
 */
function isBreakStatement(node) {
  return isNodeType(node, 'BreakStatement');
}

/**
 * @param {Object} node The node to check
 * @returns {Boolean} true if the given node is an ExpressionStatement
 */
function isExpression(node) {
  return isNodeType(node, 'ExpressionStatement');
}

/**
 * @param {Object} node The node to check
 * @returns {Boolean} true if the given node is an AssignmentStatement
 */
function isAssignment(node) {
  return isNodeType(node, 'AssignmentExpression');
}

/**
 * @param {Object} node The node to check
 * @returns {Boolean} true if the given node is an UpdateExpression
 */
function isUpdate(node) {
  return isNodeType(node, 'UpdateExpression');
}

/**
 * @param {Object} node The node to check
 * @returns {Boolean} true if the given node is an MemberExpression
 */
function isMember(node) {
  return isNodeType(node, 'MemberExpression');
}

/**
 * @param {Object} node The node to check
 * @returns {Boolean} true if the given node is an IfStatement
 */
function isIfStatement(node) {
  return isNodeType(node, 'IfStatement');
}

/**
 * @param {Object} node The node to check
 * @returns {Boolean} true if the given node is a type of loop statement
 */
function isLoopStatement(node) {
  return ['ForStatement', 'WhileStatement', 'DoWhileStatement'].indexOf(node['type']) >= 0;
}

/**
 * Determine whether a given node is a child of another given node
 * @param {Object} potentialChild AST Node to be tested for child
 * @param {Object} potentialParent AST Node to be tested for parent
 * @returns {Bool} true if potentialChild is indeed a child of potentialParent, false otherwise
 */
function isAChildOf(potentialChild, potentialParent) {
  throwIfInvalidNode(potentialChild);
  throwIfInvalidNode(potentialParen);

  // Note that if the start or end pos of both nodes is same,
  // then they're considered to be the same node and DON'T have parent-child relationship.
  return potentialChild.start > potentialParent.start && potentialChild.end < potentialParent.end;
}

function getSourceIndices(node) {
  return node.src
    .split(':')
    .map(val => parseInt(val))
    .slice(0, 2);
}

function getNodeSources(node, source) {
  const [start, len] = getSourceIndices(node);
  return [start, len, source.slice(start, start + len)];
}

/**
 * Search for the node to satisfy a specified predicate
 * @param {Object} node The AST Node to start search from
 * @returns {Object} nodeResult Result of the search
 */
function getNode(node, predicate) {
  throwIfInvalidNode(node);
  return find(node.nodes, predicate);
}

function getNodes(node, predicate) {
  throwIfInvalidNode(node);
  return node.nodes.filter(predicate);
}

function getImportDirectives(node) {
  return getNodes(node, isImportDirective);
}

function getPragmaDirectives(node) {
  return getNodes(node, isPragmaDirective);
}

function getVarDeclarations(node) {
  return getNodes(node, isVarDeclaration);
}

function getContracts(node) {
  return getNodes(node, isContractType);
}

/**
 * Search for the constructor node
 * @param {Object} node The AST Node to start search from
 * @returns {Object} nodeResult Result of the search
 */
function getConstructor(node) {
  return getNode(node, ['kind', 'constructor']);
}

/**
 * Search for the contract node by name
 * @param {Object} node The AST Node to start search from
 * @returns {Object} nodeResult Result of the search
 */
function getContract(node, contractName) {
  return getNode(node, ['name', contractName]);
}

function getContractById(node, id) {
  return getNode(node, ['id', id]);
}

/**
 * Get the parent node of the specified node
 * @param {Object} node The AST Node to retrieve the parent of
 * @returns {Object} nodeParent Parent node of the given node
 */
function getParent(node) {
  throwIfInvalidNode(node);
  return node.parent;
}

/**
 * Retrieve the line number on which the code for provided node STARTS
 * @param {Object} node The AST Node to retrieve the line number of
 * @returns {Integer} lineNumber Line number of code of the specified node. (LINES BEGIN FROM 1)
 */
function getLine(node, sourceCodeText) {
  throwIfInvalidNode(node, 'getLine');

  let newLineCharsBefore = sourceCodeText.slice(0, node.start).match(/\n/g);

  return (newLineCharsBefore ? newLineCharsBefore.length : 0) + 1;
}

/**
 * Retrieve the column number of the first character of the given node
 * @param {Object} node The AST Node to retrieve the column number of
 * @returns {Integer} columnNumber Column number of code of the specified node (COLUMNS BEGIN FROM 0)
 */
function getColumn(node, sourceCodeText) {
  throwIfInvalidNode(node);

  //start looking from sourceCodeText [node.start] and stop upon encountering the first linebreak character
  for (let i = node.start; i >= 0; i--) {
    if (sourceCodeText[i] === '\n') {
      return node.start - i - 1;
    }
  }

  return node.start;
}

/**
 * Retrieve the line number on which the code for provided node ENDS
 * @param {Object} node The AST Node to retrieve the line number of
 * @returns {int} lineNumber Line number of code ending of the specified node. (LINES BEGIN FROM 1)
 */
function getEndingLine(node, sourceCodeText) {
  throwIfInvalidNode(node);

  let newLineCharsBefore = sourceCodeText.slice(0, node.end).match(/\n/g);

  return (newLineCharsBefore ? newLineCharsBefore.length : 0) + 1;
}

/**
 * Retrieve the column number of the last character that is part of the given node
 * @param {Object} node The AST Node to retrieve the ending column number of
 * @returns {Integer} columnNumber Column number of last char of the specified node (COLUMNS BEGIN FROM 0)
 */
function getEndingColumn(node, sourceCodeText) {
  throwIfInvalidNode(node);

  //start looking from 1 character before node.start and stop upon encountering the first linebreak character
  for (let i = node.end - 1; i >= 0; i--) {
    if (sourceCodeText[i] === '\n') {
      return node.end - i - 2;
    }
  }

  return node.end - 1;
}

module.exports = {
  isASTNode,
  getNode,
  getConstructor,
  getContract,
  getContractById,
  getSourceIndices,
  getNodeSources,
  isPragmaDirective,
  isImportDirective,
  getImportDirectives,
  getPragmaDirectives,
  getVarDeclarations,
  getContracts,
  idModifierInvocation,
  isContract,
  isInterface,
};
