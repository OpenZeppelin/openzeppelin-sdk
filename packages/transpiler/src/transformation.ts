export interface Transformation {
  start: number;
  end: number;
  text: string;
}
