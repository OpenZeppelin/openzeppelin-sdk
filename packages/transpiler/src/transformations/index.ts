export { appendDirective } from './append-directive';
export { prependBaseClass } from './prepend-base-class';
export { transformParentsNames } from './transform-parents';
export { transformContractName } from './transform-contract-name';
export { purgeVarInits } from './purge-var-inits';
export { transformConstructor } from './transform-constructor';
export { purgeExceptContracts } from './purge-contracts';
export { fixImportDirectives } from './fix-import-directives';
