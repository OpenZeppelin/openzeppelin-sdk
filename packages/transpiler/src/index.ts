import flatten from 'lodash.flatten';

import { getContract, isContract, throwIfInvalidNode } from './solc/ast-utils';
import { transpile } from './transpiler';
import {
  transformConstructor,
  transformContractName,
  appendDirective,
  prependBaseClass,
  purgeExceptContracts,
  transformParentsNames,
  fixImportDirectives,
  purgeVarInits,
} from './transformations/index';
import { getInheritanceChain } from './solc/get-inheritance-chain';
import { Artifact } from './solc/artifact';
import { Transformation } from './transformation';

export interface OutputFile {
  fileName: string;
  source: string;
  path: string;
  contracts: string[];
}

interface FileTransformation {
  transformations: Transformation[];
  source: string;
}

export function transpileContracts(contracts: string[], artifacts: Artifact[]): OutputFile[] {
  // check that we have valid ast tree
  for (const art of artifacts) {
    throwIfInvalidNode(art.ast);
  }

  // create contract name | id to artifact map for quick access to artifacts
  const contractsToArtifactsMap = artifacts.reduce<Record<string | number, Artifact>>((acc, art) => {
    acc[art.contractName] = art;
    const contract = getContract(art);
    acc[contract.id] = art;
    return acc;
  }, {});

  // build a list of all contracts to transpile
  const contractsToTranspile = [
    ...new Set(flatten(contracts.map(contract => getInheritanceChain(contract, contractsToArtifactsMap)))),
  ].filter(art => {
    const contractNode = getContract(art);
    return isContract(contractNode);
  });

  // build a array of transformations per Solidity file
  const fileTrans = contractsToTranspile.reduce<Record<string, FileTransformation>>((acc, art) => {
    const contractName = art.contractName;

    const source = art.source;

    const contractNode = getContract(art);

    if (!acc[art.fileName]) {
      const directive = `\nimport "@openzeppelin/upgrades/contracts/Initializable.sol";`;

      acc[art.fileName] = {
        transformations: [
          appendDirective(art.ast, directive),
          ...fixImportDirectives(art, artifacts, contractsToTranspile),
          ...purgeExceptContracts(art.ast, contractsToTranspile),
        ],
        source: '',
      };
    }

    acc[art.fileName].transformations = [
      ...acc[art.fileName].transformations,
      prependBaseClass(contractNode, source, 'Initializable'),
      ...transformParentsNames(contractNode, source, contractsToTranspile),
      ...transformConstructor(contractNode, source, contractsToTranspile, contractsToArtifactsMap),
      ...purgeVarInits(contractNode, source),
      transformContractName(contractNode, source, `${contractName}Upgradable`),
    ];

    return acc;
  }, {});

  // build a final array of files to return
  return contractsToTranspile.reduce<OutputFile[]>((acc, art) => {
    const contractName = art.contractName;

    const source = art.source;

    const fileTran = fileTrans[art.fileName];
    if (!fileTran.source) {
      fileTran.source = transpile(source, fileTran.transformations);
    }
    const entry = acc.find(o => o.fileName === art.fileName);
    if (!entry) {
      const path = art.sourcePath.replace('.sol', 'Upgradable.sol');
      let patchedFilePath = path;
      if (path.startsWith('contracts')) {
        patchedFilePath = path.replace('contracts/', '');
      }
      patchedFilePath = `./contracts/__upgradable__/${patchedFilePath}`;
      acc.push({
        source: fileTran.source,
        path: patchedFilePath,
        fileName: art.fileName,
        contracts: [contractName],
      });
    } else {
      entry.contracts.push(contractName);
    }
    return acc;
  }, []);
}
