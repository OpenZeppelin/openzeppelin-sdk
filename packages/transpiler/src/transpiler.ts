import { Transformation } from './transformation';

export function transpile(source: string, transformations: Transformation[]): string {
  let cursor = 0;

  const sorted = transformations.sort((a, b) => {
    return a.start - b.start;
  });

  for (let i = 0; i < sorted.length - 1; i++) {
    if (sorted[i].end > sorted[i + 1].start)
      throw new Error(
        `Transformations ${sorted[i].start}:${sorted[i].end}:${sorted[i].text.slice(20)}... and ${
          sorted[i + 1].start
        }:${sorted[i + 1].end}:${sorted[i + 1].text.slice(20)}... overlap over the source file`,
      );
  }

  let transpiledCode = sorted.reduce((output, trans) => {
    const { start, end, text } = trans;
    output += source.slice(cursor, start);
    output += text;
    cursor = end;
    return output;
  }, '');

  transpiledCode += source.slice(cursor);
  return transpiledCode;
}
