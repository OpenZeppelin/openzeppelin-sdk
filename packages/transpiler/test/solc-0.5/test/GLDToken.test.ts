import { transpileContracts } from '../../../src/';
import fs from 'fs-extra';

const artifacts = fs.readdirSync('./build/contracts/').map(file => {
  return JSON.parse(fs.readFileSync(`./build/contracts/${file}`).toString());
});

describe('GLDToken contract', (): void => {
  it('is converted to a valid GLDTokenUpgradable contract', (): void => {
    const [GLDTokenFile, ERC20DetailedFile, ERC20File, ContextFile] = transpileContracts(['GLDToken'], artifacts);

    expect(GLDTokenFile.source).toMatchSnapshot();
    expect(GLDTokenFile.fileName).toBe('GLDToken.sol');
    expect(GLDTokenFile.path).toBe('./contracts/__upgradable__/GLDTokenUpgradable.sol');
    expect(GLDTokenFile.contracts).toEqual(['GLDToken']);

    expect(ERC20DetailedFile.source).toMatchSnapshot();
    expect(ERC20DetailedFile.fileName).toBe('ERC20Detailed.sol');
    expect(ERC20DetailedFile.path).toBe(
      './contracts/__upgradable__/@openzeppelin/contracts/token/ERC20/ERC20DetailedUpgradable.sol',
    );
    expect(ERC20DetailedFile.contracts).toEqual(['ERC20Detailed']);

    expect(ERC20File.source).toMatchSnapshot();
    expect(ERC20File.fileName).toBe('ERC20.sol');
    expect(ERC20File.path).toBe('./contracts/__upgradable__/@openzeppelin/contracts/token/ERC20/ERC20Upgradable.sol');
    expect(ERC20File.contracts).toEqual(['ERC20']);

    expect(ContextFile.source).toMatchSnapshot();
    expect(ContextFile.fileName).toBe('Context.sol');
    expect(ContextFile.path).toBe('./contracts/__upgradable__/@openzeppelin/contracts/GSN/ContextUpgradable.sol');
    expect(ContextFile.contracts).toEqual(['Context']);
  });
});
