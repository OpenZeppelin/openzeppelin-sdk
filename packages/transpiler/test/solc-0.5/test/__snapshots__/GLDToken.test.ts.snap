// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GLDToken contract is converted to a valid GLDTokenUpgradable contract 1`] = `
"pragma solidity ^0.5.0;

import \\"./@openzeppelin/contracts/token/ERC20/ERC20Upgradable.sol\\";
import \\"./@openzeppelin/contracts/token/ERC20/ERC20DetailedUpgradable.sol\\";
import \\"@openzeppelin/upgrades/contracts/Initializable.sol\\";

contract GLDTokenUpgradable is Initializable, ERC20Upgradable, ERC20DetailedUpgradable {
    function initialize(uint256 initialSupply) external initializer {
        __init(true, initialSupply);
    }

    function __init(bool callChain, uint256 initialSupply) internal {
        if(callChain) {
            ContextUpgradable.__init(false);
            ERC20Upgradable.__init(false);
            ERC20DetailedUpgradable.__init(false, \\"Gold\\", \\"GLD\\", 18);
        }
        
        
        _mint(msg.sender, initialSupply);
    
    }

    
}
"
`;

exports[`GLDToken contract is converted to a valid GLDTokenUpgradable contract 2`] = `
"pragma solidity ^0.5.0;

import \\"@openzeppelin/contracts/token/ERC20/IERC20.sol\\";
import \\"@openzeppelin/upgrades/contracts/Initializable.sol\\";

/**
 * @dev Optional functions from the ERC20 standard.
 */
contract ERC20DetailedUpgradable is Initializable, IERC20 {

    function __init(bool callChain, string memory name, string memory symbol, uint8 decimals) internal {
        
        
        
        _name = name;
        _symbol = symbol;
        _decimals = decimals;
    
    }

    string private _name;
    string private _symbol;
    uint8 private _decimals;

    /**
     * @dev Sets the values for \`name\`, \`symbol\`, and \`decimals\`. All three of
     * these values are immutable: they can only be set once during
     * construction.
     */
    

    /**
     * @dev Returns the name of the token.
     */
    function name() public view returns (string memory) {
        return _name;
    }

    /**
     * @dev Returns the symbol of the token, usually a shorter version of the
     * name.
     */
    function symbol() public view returns (string memory) {
        return _symbol;
    }

    /**
     * @dev Returns the number of decimals used to get its user representation.
     * For example, if \`decimals\` equals \`2\`, a balance of \`505\` tokens should
     * be displayed to a user as \`5,05\` (\`505 / 10 ** 2\`).
     *
     * Tokens usually opt for a value of 18, imitating the relationship between
     * Ether and Wei.
     *
     * NOTE: This information is only used for _display_ purposes: it in
     * no way affects any of the arithmetic of the contract, including
     * {IERC20-balanceOf} and {IERC20-transfer}.
     */
    function decimals() public view returns (uint8) {
        return _decimals;
    }
}
"
`;

exports[`GLDToken contract is converted to a valid GLDTokenUpgradable contract 3`] = `
"pragma solidity ^0.5.0;

import \\"../../GSN/ContextUpgradable.sol\\";
import \\"@openzeppelin/contracts/token/ERC20/IERC20.sol\\";
import \\"@openzeppelin/contracts/math/SafeMath.sol\\";
import \\"@openzeppelin/upgrades/contracts/Initializable.sol\\";

/**
 * @dev Implementation of the {IERC20} interface.
 *
 * This implementation is agnostic to the way tokens are created. This means
 * that a supply mechanism has to be added in a derived contract using {_mint}.
 * For a generic mechanism see {ERC20Mintable}.
 *
 * TIP: For a detailed writeup see our guide
 * https://forum.zeppelin.solutions/t/how-to-implement-erc20-supply-mechanisms/226[How
 * to implement supply mechanisms].
 *
 * We have followed general OpenZeppelin guidelines: functions revert instead
 * of returning \`false\` on failure. This behavior is nonetheless conventional
 * and does not conflict with the expectations of ERC20 applications.
 *
 * Additionally, an {Approval} event is emitted on calls to {transferFrom}.
 * This allows applications to reconstruct the allowance for all accounts just
 * by listening to said events. Other implementations of the EIP may not emit
 * these events, as it isn't required by the specification.
 *
 * Finally, the non-standard {decreaseAllowance} and {increaseAllowance}
 * functions have been added to mitigate the well-known issues around setting
 * allowances. See {IERC20-approve}.
 */
contract ERC20Upgradable is Initializable, ContextUpgradable, IERC20 {
    function initialize() external initializer {
        __init(true);
    }

    function __init(bool callChain) internal {
        if(callChain) {
            ContextUpgradable.__init(false);
        }
        
        
    }

    using SafeMath for uint256;

    mapping (address => uint256) private _balances;

    mapping (address => mapping (address => uint256)) private _allowances;

    uint256 private _totalSupply;

    /**
     * @dev See {IERC20-totalSupply}.
     */
    function totalSupply() public view returns (uint256) {
        return _totalSupply;
    }

    /**
     * @dev See {IERC20-balanceOf}.
     */
    function balanceOf(address account) public view returns (uint256) {
        return _balances[account];
    }

    /**
     * @dev See {IERC20-transfer}.
     *
     * Requirements:
     *
     * - \`recipient\` cannot be the zero address.
     * - the caller must have a balance of at least \`amount\`.
     */
    function transfer(address recipient, uint256 amount) public returns (bool) {
        _transfer(_msgSender(), recipient, amount);
        return true;
    }

    /**
     * @dev See {IERC20-allowance}.
     */
    function allowance(address owner, address spender) public view returns (uint256) {
        return _allowances[owner][spender];
    }

    /**
     * @dev See {IERC20-approve}.
     *
     * Requirements:
     *
     * - \`spender\` cannot be the zero address.
     */
    function approve(address spender, uint256 amount) public returns (bool) {
        _approve(_msgSender(), spender, amount);
        return true;
    }

    /**
     * @dev See {IERC20-transferFrom}.
     *
     * Emits an {Approval} event indicating the updated allowance. This is not
     * required by the EIP. See the note at the beginning of {ERC20};
     *
     * Requirements:
     * - \`sender\` and \`recipient\` cannot be the zero address.
     * - \`sender\` must have a balance of at least \`amount\`.
     * - the caller must have allowance for \`sender\`'s tokens of at least
     * \`amount\`.
     */
    function transferFrom(address sender, address recipient, uint256 amount) public returns (bool) {
        _transfer(sender, recipient, amount);
        _approve(sender, _msgSender(), _allowances[sender][_msgSender()].sub(amount, \\"ERC20: transfer amount exceeds allowance\\"));
        return true;
    }

    /**
     * @dev Atomically increases the allowance granted to \`spender\` by the caller.
     *
     * This is an alternative to {approve} that can be used as a mitigation for
     * problems described in {IERC20-approve}.
     *
     * Emits an {Approval} event indicating the updated allowance.
     *
     * Requirements:
     *
     * - \`spender\` cannot be the zero address.
     */
    function increaseAllowance(address spender, uint256 addedValue) public returns (bool) {
        _approve(_msgSender(), spender, _allowances[_msgSender()][spender].add(addedValue));
        return true;
    }

    /**
     * @dev Atomically decreases the allowance granted to \`spender\` by the caller.
     *
     * This is an alternative to {approve} that can be used as a mitigation for
     * problems described in {IERC20-approve}.
     *
     * Emits an {Approval} event indicating the updated allowance.
     *
     * Requirements:
     *
     * - \`spender\` cannot be the zero address.
     * - \`spender\` must have allowance for the caller of at least
     * \`subtractedValue\`.
     */
    function decreaseAllowance(address spender, uint256 subtractedValue) public returns (bool) {
        _approve(_msgSender(), spender, _allowances[_msgSender()][spender].sub(subtractedValue, \\"ERC20: decreased allowance below zero\\"));
        return true;
    }

    /**
     * @dev Moves tokens \`amount\` from \`sender\` to \`recipient\`.
     *
     * This is internal function is equivalent to {transfer}, and can be used to
     * e.g. implement automatic token fees, slashing mechanisms, etc.
     *
     * Emits a {Transfer} event.
     *
     * Requirements:
     *
     * - \`sender\` cannot be the zero address.
     * - \`recipient\` cannot be the zero address.
     * - \`sender\` must have a balance of at least \`amount\`.
     */
    function _transfer(address sender, address recipient, uint256 amount) internal {
        require(sender != address(0), \\"ERC20: transfer from the zero address\\");
        require(recipient != address(0), \\"ERC20: transfer to the zero address\\");

        _balances[sender] = _balances[sender].sub(amount, \\"ERC20: transfer amount exceeds balance\\");
        _balances[recipient] = _balances[recipient].add(amount);
        emit Transfer(sender, recipient, amount);
    }

    /** @dev Creates \`amount\` tokens and assigns them to \`account\`, increasing
     * the total supply.
     *
     * Emits a {Transfer} event with \`from\` set to the zero address.
     *
     * Requirements
     *
     * - \`to\` cannot be the zero address.
     */
    function _mint(address account, uint256 amount) internal {
        require(account != address(0), \\"ERC20: mint to the zero address\\");

        _totalSupply = _totalSupply.add(amount);
        _balances[account] = _balances[account].add(amount);
        emit Transfer(address(0), account, amount);
    }

     /**
     * @dev Destroys \`amount\` tokens from \`account\`, reducing the
     * total supply.
     *
     * Emits a {Transfer} event with \`to\` set to the zero address.
     *
     * Requirements
     *
     * - \`account\` cannot be the zero address.
     * - \`account\` must have at least \`amount\` tokens.
     */
    function _burn(address account, uint256 amount) internal {
        require(account != address(0), \\"ERC20: burn from the zero address\\");

        _balances[account] = _balances[account].sub(amount, \\"ERC20: burn amount exceeds balance\\");
        _totalSupply = _totalSupply.sub(amount);
        emit Transfer(account, address(0), amount);
    }

    /**
     * @dev Sets \`amount\` as the allowance of \`spender\` over the \`owner\`s tokens.
     *
     * This is internal function is equivalent to \`approve\`, and can be used to
     * e.g. set automatic allowances for certain subsystems, etc.
     *
     * Emits an {Approval} event.
     *
     * Requirements:
     *
     * - \`owner\` cannot be the zero address.
     * - \`spender\` cannot be the zero address.
     */
    function _approve(address owner, address spender, uint256 amount) internal {
        require(owner != address(0), \\"ERC20: approve from the zero address\\");
        require(spender != address(0), \\"ERC20: approve to the zero address\\");

        _allowances[owner][spender] = amount;
        emit Approval(owner, spender, amount);
    }

    /**
     * @dev Destroys \`amount\` tokens from \`account\`.\`amount\` is then deducted
     * from the caller's allowance.
     *
     * See {_burn} and {_approve}.
     */
    function _burnFrom(address account, uint256 amount) internal {
        _burn(account, amount);
        _approve(account, _msgSender(), _allowances[account][_msgSender()].sub(amount, \\"ERC20: burn amount exceeds allowance\\"));
    }
}
"
`;

exports[`GLDToken contract is converted to a valid GLDTokenUpgradable contract 4`] = `
"pragma solidity ^0.5.0;
import \\"@openzeppelin/upgrades/contracts/Initializable.sol\\";

/*
 * @dev Provides information about the current execution context, including the
 * sender of the transaction and its data. While these are generally available
 * via msg.sender and msg.data, they should not be accessed in such a direct
 * manner, since when dealing with GSN meta-transactions the account sending and
 * paying for execution may not be the actual sender (as far as an application
 * is concerned).
 *
 * This contract is only required for intermediate, library-like contracts.
 */
contract ContextUpgradable is Initializable {
    function initialize() external initializer {
        __init(true);
    }

    function __init(bool callChain) internal {
        
        
         
    }

    // Empty internal constructor, to prevent people from mistakenly deploying
    // an instance of this contract, which should be used via inheritance.
    
    // solhint-disable-previous-line no-empty-blocks

    function _msgSender() internal view returns (address payable) {
        return msg.sender;
    }

    function _msgData() internal view returns (bytes memory) {
        this; // silence state mutability warning without generating bytecode - see https://github.com/ethereum/solidity/issues/2691
        return msg.data;
    }
}
"
`;
